### PR TITLE
Common user Task methods cannot return null

### PR DESCRIPTION
**Pull request by boblangley**
Closes https://github.com/Particular/NServiceBus/issues/3533

Added warnings to appropriate doco pages to make it explicit that returning `null` doesn't work for version 6.

/cc: @danielmarbach

### CODE REVIEW

**Comment by SimonCropp** [comment]
@danielmarbach can u review the async stuff here

**Comment by danielmarbach** [comment]
Not before Thursday :(

On Tuesday, March 8, 2016, Simon Cropp <notifications@github.com> wrote:

> @danielmarbach <https://github.com/danielmarbach> can u review the async
> stuff here
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Particular/docs.particular.net/pull/1288#issuecomment-193995762>
> .
>


-- 
*Don't miss the Async/Await Webinar Series*
*• Learn more <http://go.particular.net/l/27442/2016-02-15/681bjl> →*

File: nservicebus/sagas/saga-finding.md [hunk 1]
```diff
@@ -21,7 +21,8 @@ snippet:nhibernate-saga-finder
 
 snippet:ravendb-saga-finder
 
+Starting with version 6 all handlers must return a `Task`, a completed `Task` or be marked `async`. Returning null will throw an `InvalidOperationException`.
 
-You can have as many of these classes as you want for a given saga or message type. If a saga can't be found, return null, and if the saga specifies that it is to be started for that message type, NServiceBus will know that a new saga instance is to be created. The above example uses NServiceBus extension for NHibernate that allows both framework and user code to share the same NHibernate session. Similar extension point exists for RavenDB.
+You can have as many of these classes as you want for a given saga or message type. If a saga can't be found, return null (or `Task.FromResult(null)` starting with version 6), and if the saga specifies that it is to be started for that message type, NServiceBus will know that a new saga instance is to be created. The above example uses NServiceBus extension for NHibernate that allows both framework and user code to share the same NHibernate session. Similar extension point exists for RavenDB.
```
Comment (mikeminutillo) [inline nservicebus/sagas/saga-finding.md]: Is this an opportunity to return something more intentional like `SagaFinder.NotFound()`?

File: nservicebus/lifecycle/iwanttorunwhenbusstartsandstops.md [hunk 2]
```diff
@@ -28,18 +28,23 @@ Instances are:
 
 Exceptions thrown in the constructors of instances of `IWantToRunWhenBusStartsAndStops` are unhandled by NServiceBus. These will bubble up to the code that starts the the bus.
 
-Once created `Start()` is called on each instance asynchronously without awaiting its completion. Each `Start()` call is made on a on the same thread which started the bus. Instances of `IWantToRunWhenBusStartsAndStops` are kept internally to be stopped when the bus is stopped.
+Once created `Start()` is called on each instance asynchronously without awaiting its completion. Each `Start()` call is made on a on the same thread which started the bus. Instances of `IWantToRunWhenBusStartsAndStops` are kept internally to be stopped when the bus is stopped.
```
Comment (mikeminutillo) [inline nservicebus/lifecycle/iwanttorunwhenbusstartsandstops.md]: > Each `Start()` call is made **on a on the** same thread which started the bus.

File: nservicebus/sagas/saga-finding.md (hunk 1 above)
Comment (boblangley) [inline nservicebus/sagas/saga-finding.md]: I think `null` is appropriate, but perhaps the name could be better similar to `SingleOrDefault`?

File: nservicebus/sagas/timeouts.md [hunk 3]
```diff
@@ -22,6 +22,7 @@ NOTE: If the saga does not request a timeout then the corresponding timeout meth
 
 WARNING: Don't assume that other messages haven't arrived in the meantime. If required a Saga can store boolean flags in the SagaData and then check these flags to confirm a given timeout message should be processed based on the current sate.
 
+Starting with version 6 all timeout handlers must return a `Task`, a completed `Task` or be marked `async`. Returning null will throw an `InvalidOperationException`.
```
Comment (danielmarbach) [inline nservicebus/sagas/timeouts.md]: Should it be `null`?

File: nservicebus/sagas/saga-not-found.md [hunk 4]
```diff
@@ -13,4 +13,6 @@ snippet:saga-not-found
 
 Note that the message will be considered successfully processed and sent to the audit queue even if no saga was found. If you want the message to end up in the error queue just throw an exception from your `IHandleSagaNotFound` implementation.
 
+Starting with version 6 all handlers must return a `Task`, a completed `Task` or be marked `async`. Returning null will throw an `InvalidOperationException`.
```
Comment (danielmarbach) [inline nservicebus/sagas/saga-not-found.md]: Should it be `null`?

File: nservicebus/sagas/saga-finding.md (hunk 1 above)
Comment (danielmarbach) [inline nservicebus/sagas/saga-finding.md]: >  If a saga can't be found, return null (or `Task.FromResult(null)` starting with version 6), and if the saga specifies that it is to be started for that message type

How about

> For NServiceBus version 5 and below return `null` if a saga can't be found. Starting with version 6, return `null` when the method is marked `async` or `Task.FromResult(null)` when the method is not marked as `async`.

File: nservicebus/lifecycle/iwanttorunwhenbusstartsandstops.md [hunk 5]
```diff
@@ -28,18 +28,23 @@ Instances are:
...
 
+Start must return a `Task`, a completed `Task` or be marked `async` otherwise an `InvalidOperationException` will be thrown.
+
 Exceptions raised from the `Start()` method will cause the startup process to be aborted and the exception is raised to the caller.
 
 NOTE: The call to `IStartableEndpoint.Start()` will not return before all instances of `IWantToRunWhenBusStartsAndStops.Start()` are completed.
 
-When the Bus is disposed, all instances of `IWantToRunWhenBusStartsAndStops` are stopped by calling their `Stop()` method asynchronously but not on its dedicated thread. Each call to `Stop()` happens on the same thread which disposes the bus. Any exceptions thrown by a call to `Stop` will be logged at the Fatal level.
+When the Bus is disposed, all instances of `IWantToRunWhenBusStartsAndStops` are stopped by calling their `Stop()` method asynchronously but not on its dedicated thread. Each call to `Stop()` happens on the same thread which disposes the bus.
```
Comment (danielmarbach) [inline nservicebus/lifecycle/iwanttorunwhenbusstartsandstops.md]: > which disposes the bus

which stops the bus.

File: nservicebus/sagas/saga-finding.md (hunk 1 above)
Comment (mikeminutillo) [inline nservicebus/sagas/saga-finding.md]: A few minor changes:
- As these are the NServiceBus docs, we usually don't specify NServiceBus when referring to specific versions
- Version should be capitalized.
- I tried to simplify it a bit.

> If a saga can't be found return `null`. In Version 6 and above, if the method is not marked `async`, return `Task.FromResult(null)`.

Comment (danielmarbach) [inline nservicebus/sagas/saga-finding.md]: :+1:

**Comment by boblangley** [comment]
Ready for review @danielmarbach / @Particular/docs-maintainers

File: nservicebus/lifecycle/iwanttorunwhenbusstartsandstops.md [hunk 6]
```diff
@@ -32,14 +32,19 @@ Once created `Start()` is called on each instance asynchronously without awaitin
 
 NOTE: Since `Start()` executes all `IWantToRunWhenBusStartsAndStops` asynchronously but not on its dedicated thread. It is the responsibility of the implementing class to execute its operations in parallel if needed (i.ex. for CPU bound work). Failure to do so will prevent the bus from being started.
 
+Start must return a `Task`, a completed `Task` or be marked `async`.
```
Comment (danielmarbach) [inline nservicebus/lifecycle/iwanttorunwhenbusstartsandstops.md]: `Start()` ?

File: nservicebus/lifecycle/iwanttorunwhenbusstartsandstops.md [hunk 7]
```diff
@@ -32,14 +32,19 @@ Once created `Start()` is called on each instance asynchronously without awaitin
...
+
 Exceptions raised from the `Start()` method will cause the startup process to be aborted and the exception is raised to the caller.
 
 NOTE: The call to `IStartableEndpoint.Start()` will not return before all instances of `IWantToRunWhenBusStartsAndStops.Start()` are completed.
 
-When the Bus is disposed, all instances of `IWantToRunWhenBusStartsAndStops` are stopped by calling their `Stop()` method asynchronously but not on its dedicated thread. Each call to `Stop()` happens on the same thread which disposes the bus. Any exceptions thrown by a call to `Stop` will be logged at the Fatal level.
+When the Bus is disposed, all instances of `IWantToRunWhenBusStartsAndStops` are stopped by calling their `Stop()` method asynchronously but not on its dedicated thread. Each call to `Stop()` happens on the same thread which stops the bus. 
 
-NOTE: `Stop()` will wait for any outstanding instances of `Start()` to complete. It is the responsibility of the implementing class to execute its operations in parallel if needed (i.ex. for CPU bound work). Failure to do so will prevent the bus from being disposed.
+Stop must return a `Task`, a completed `Task` or be marked `async`.
```
Comment (danielmarbach) [inline nservicebus/lifecycle/iwanttorunwhenbusstartsandstops.md]: `Stop()` ?

**Comment by danielmarbach** [comment]
Two comments. Apart from that :+1: But don't trust me. It is late

**Comment by boblangley** [comment]
@Particular/docs-maintainers ready to merge on your approval

**Pull request merged**